search-service/src/query_representation/ultimate/mod.rs

```diff
@@ -0,0 +1,248 @@
+/*
+    Module responsible for creating the final (SQL) representation
+    of a query. It receives the command created in the intermediary representation
+    and the projection coming from the initial representation.
+*/
+
+use anyhow::Error;
+
+use crate::query_representation::intermediary::{
+    get_command_attributes, simple_command::Operator, Command,
+};
+
+pub fn command_to_query(projection: Vec<String>, command: &Command) -> Result<String, Error> {
+    let mut attributes_needed = projection.clone();
+    attributes_needed.extend(get_command_attributes(command));
+
+    let tables_needed = get_tables_needed(attributes_needed)?;
+
+    let _select_query = create_select_query(projection);
+
+    let _from_query = create_from_query(tables_needed);
+
+    let _where_query = create_where_query(command, true)?;
+
+    // let final_query = [select_query, from_query, where_query].join("\n");
+    let final_query = "Command to query not implemented yet".to_string();
+
+    Ok(final_query)
+}
+
+fn get_tables_needed(_attributes: Vec<String>) -> Result<Vec<String>, Error> {
+    /* TODO: Given a list of attributes, return the tables needed to join all of them
+    in one query (using src/relational/tableSearch) */
+
+    let tables: Vec<String> = vec![];
+
+    Ok(tables)
+}
+
+fn create_select_query(projection: Vec<String>) -> String {
+    let mut select_query = "SELECT ".to_owned();
+    let len = projection.len();
+
+    for (idx, column) in projection.iter().enumerate() {
+        select_query.push_str(column);
+        if idx != len - 1 {
+            select_query.push_str(", ");
+        }
+    }
+
+    select_query
+}
+
+fn create_from_query(tables: Vec<String>) -> String {
+    let mut from_query = "FROM ".to_owned();
+    let len = tables.len();
+
+    for (idx, table) in tables.iter().enumerate() {
+        from_query.push_str(table);
+        if idx != len - 1 {
+            from_query.push_str(", ");
+        }
+    }
+
+    from_query
+}
+
+fn create_where_query(command: &Command, initial_call: bool) -> Result<String, Error> {
+    let mut where_query = "".to_owned();
+
+    if initial_call {
+        where_query.push_str("WHERE ")
+        // TODO: Add JOIN filters here
+    };
+
+    match command {
+        Command::CompositeCommand(composite_command) => {
+            let nested_commands = &composite_command.commands;
+
+            if !initial_call {
+                where_query.push_str("(")
+            }
+
+            where_query.push_str(&create_where_query(&nested_commands[0], false)?);
+            where_query.push_str(&composite_command.operation.to_string());
```

Comment:
Penso que deveríamos mudar operation para LogicalOperator. Pra mim não faz muito sentido chamar AND/OR de operação...

Reply:
Commit 9982a568824c74519e3c9c11446319421d85a583